### PR TITLE
Fix rolling average gas price

### LINKED ISSUE #477: Running an Edge network with 0 transaction fee
### Description

After running the polygon edge network with below `config.json` and `genesis.json` file, when I query the validator node with the `eth_gasPrice` RPC method it sometimes returns zero and sometimes non-zero values. And when the value is non-zero, I am unable to make the transaction with 0 `gasPrice` as it throws the insufficient funds error.

#### `config.json`
```
{
  "data_dir": "/data/chain",
  "chain_config": "/data/chain/genesis.json",
  "block_time_s": 15,
  "jsonrpc_addr": "0.0.0.0:8545",
  "grpc_addr": "0.0.0.0:9632",
  "seal": true,
  "network": {
    "libp2p_addr": "0.0.0.0:1478",
    "dns_addr": "dns/staging-luis-polygon-edge-01-val-01-9721p.settlemint.com"
  },
  "txpool": {
    "price_limit": 0
  },
  "telemetry": {
    "prometheus_addr": "0.0.0.0:9545"
  }
}
```

#### `genesis.json`
```
{
  "name": "staging-luis-polygon-edge-01-ace8",
  "genesis": {
    "nonce": "0x0000000000000000",
    "timestamp": "0x0",
    "gasLimit": "9007199254740991",
    "difficulty": "0x1",
    "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "coinbase": "0x0000000000000000000000000000000000000000",
    "alloc": {
      "0x227FBE30d0e9196c889785f539F3C93523420163": {
        "balance": "0x446c3b15f9926687d2c40534fdb564000000000000"
      }
    },
    "number": "0x0",
    "gasUsed": "0x70000",
    "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "extraData": "0x0000000000000000000000000000000000000000000000000000000000000000d8d594b34c88847ea50bf8304f89ba5aab5b844785fd8680c0"
  },
  "params": {
    "forks": {
      "homestead": 0,
      "byzantium": 0,
      "constantinople": 0,
      "petersburg": 0,
      "istanbul": 0,
      "EIP150": 0,
      "EIP158": 0,
      "EIP155": 0
    },
    "chainID": 41510,
    "engine": {
      "ibft": {
        "type": "PoA"
      }
    },
    "blockGasTarget": 10378291982571407000
  },
  "bootnodes": [
    "/dns/staging-luis-polygon-edge-01-val-01-9721p.settlemint.com/tcp/1478/p2p/16Uiu2HAmT7eRZ3LcfgZsoKGPPJ7fP3xjfqEXc818WnJqGjSnkCss"
  ]
}
```

And here is the response from the `eth_gasPrice` RPC method -
```
{
  "jsonrpc":"2.0"
  "id":42
  "result":"0x0"
}
```

```
{
"jsonrpc":"2.0"
"id":42
"result":"0x19e6"
}

```
### Your environment

* OS and version: Kubernetes GCP/EKS
* version of the Polygon Edge: v0.3.1
* branch that causes this issue: NA

### Steps to reproduce

After running the network, run the `eth_gasPrice` ethereum RPC method to see that the gasPrice is sometimes non-zero and is dynamic.

### Expected behaviour

We are expecting gasPrice to be static and 0, so transactions won't require a network token to make a transaction.

### Actual behaviour

gasPrice is coming dynamic with non-zero values sometimes and because of this, the transactions are failing with insufficient funds error.

### Logs

NA

### Proposed solution

NA

### PR DESCRIPTION
**Pull request by zivkovicmilos**
# Description

This PR fixes a bug with calculating the average gas price for the blockchain.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Added coverage tests for possible edge-cases.

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -67,23 +71,82 @@ type BlockResult struct {
...
+	}
+
+	// There is existing average data,
+	// use it to generate a new average
+	b.getRollingAverage(newValues, sum)
+}
+
+// getArithmeticAverage calculates and sets the arithmetic average
+// of the passed in data set
+func (b *Blockchain) getArithmeticAverage(newValues []*big.Int, sum *big.Int) {
```
Comment (Kourin1996) [inline blockchain/blockchain.go]: why get...? This method updates its own state.

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -67,23 +71,82 @@ type BlockResult struct {
...
+		oldCount   = b.gpAverage.count
+		oldAverage = b.gpAverage.price
 
-	b.averageGasPriceCount.Add(b.averageGasPriceCount, big.NewInt(1))
+		inputSetCount = big.NewInt(0).SetInt64(int64(len(newValues)))
+	)
 
-	differential := big.NewInt(0)
-	differential.Div(newValue.Sub(newValue, b.averageGasPrice), b.averageGasPriceCount)
+	// old average * (n-len(M))/n
```
Comment (Kourin1996) [inline blockchain/blockchain.go:121]: I don't think formula in comment matches code
{old_average*(n-len(M))}/n?

**Review by Kourin1996 — CHANGES_REQUESTED** [review]
Looks good. But I've left some small code improvements. Please check them

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -67,23 +71,82 @@ type BlockResult struct {
...
+
+	b.gpAverage.price = newAverage
+	b.gpAverage.count = newAverageCount
+}
+
+// getRollingAverage calculates the new average based on the
+// moving average formula:
+// new average = old average * (n-len(M))/n + (sum of values in M)/n)
+// where n is the old average data count, and M is the new data set
+func (b *Blockchain) getRollingAverage(newValues []*big.Int, sum *big.Int) {
```
Comment (Kourin1996) [inline blockchain/blockchain.go]: This is also. This methods updates its own fields.

File: blockchain/blockchain.go (hunk 1 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go]: You're completely right, it's leftover 🤦 

Fixed: 

[34498f1](https://github.com/0xPolygon/polygon-edge/pull/483/commits/34498f1e86e3847fbf343bc4b87d7ca430e20ca8)

File: blockchain/blockchain.go (hunk 3 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go]: You're completely right, it's leftover 🤦 

Fixed: 

[34498f1](https://github.com/0xPolygon/polygon-edge/pull/483/commits/34498f1e86e3847fbf343bc4b87d7ca430e20ca8)

File: blockchain/blockchain.go (hunk 2 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go:121]: It matches it:
`old average * (n-len(M))/n`:
* `n - len(M)` = A
* `old average * A` = B
* `B / n` = C
```go
	newAverage := big.NewInt(0).Div( // C
		big.NewInt(0).Mul( // B
			oldAverage,
			big.NewInt(0).Sub(oldCount, inputSetCount), // A
		),
		oldCount,
	)
```

Comment (zivkovicmilos) [reply blockchain/blockchain.go:121]: Do you want us to drop the comments? Maybe it's too convoluted to have them there

Comment (zivkovicmilos) [reply blockchain/blockchain.go:121]: cc @dbrajovic

**Review by dbrajovic — APPROVED** [review]

File: blockchain/blockchain.go (hunk 2 above)
Comment (Kourin1996) [reply blockchain/blockchain.go:121]: OK. It was my misunderstood. Sorry

**Review by Kourin1996 — APPROVED** [review]
